### PR TITLE
Add complete  test cases for SparkApplication

### PR DESCRIPTION
**Pull request by Abhay349**
### What does this PR do?

This PR adds test cases for
`SparkApplication (sparkoperator.k8s.io/v1beta2)`.

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Thanks
/assign

**Comment by XiShanYongYe-Chang** [comment]
Hi @FAUST-BENCHOU Would you like to review this PR?

**Comment by FAUST-BENCHOU** [comment]
> Hi @FAUST-BENCHOU Would you like to review this PR?

Sure!let me take a look at relative lua script first.

**Comment by FAUST-BENCHOU** [comment]
The core tests are covered, but the boundary case tests are relatively weak. Below are some of my suggestions for improvement (of course, you can extend these suggestions as well).
### AggregateStatus
1.Several status items have missing fields: Some untested items may lack fields such as executionAttempts and submissionAttempts.
2.The comparison and other logic for lastSubmissionAttemptTime and terminationTime.
https://github.com/karmada-io/karmada/blob/04aed89718dc0a5378b87601038f6ba45803c262/pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/customizations.yaml#L302-L305

### InterpretComponent
1.GPU resources related
https://github.com/karmada-io/karmada/blob/04aed89718dc0a5378b87601038f6ba45803c262/pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/customizations.yaml#L187-L194
2.some default value like ```executor.instances```  `driver.memory` `executor.memory`......
3.Pod template related
https://github.com/karmada-io/karmada/blob/04aed89718dc0a5378b87601038f6ba45803c262/pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/customizations.yaml#L84-L108

### InterpretHealth 
all in all good.maybe you can also consider scenarios where `state` is an empty string or `applicationState` exists but the `state` field is missing.

### InterpretStatus (ReflectStatus)
good but maybe u can consider that when applicationState, driverInfo, executorState, etc. are nil, a copy will be made.

@Abhay349 Thanks for you awesome job!

**Comment by FAUST-BENCHOU** [comment]
@XiShanYongYe-Chang This is just my opinion; any feedback is welcome.🙃

**Comment by Ferryx349** [comment]
@FAUST-BENCHOU Thanks for review..
let me look on those issues.

**Comment by Ferryx349** [comment]
Thanks a lot for the review and suggestions 
I’m a beginner with these interpreter tests, so this feedback really helps.
I just wanted to confirm, for cases where some status fields are missing, should we add explicit tests for those scenarios or treat them as default values?

I’ll update the tests accordingly. Thanks again!

**Comment by FAUST-BENCHOU** [comment]
> Thanks a lot for the review and suggestions I’m a beginner with these interpreter tests, so this feedback really helps. I just wanted to confirm, for cases where some status fields are missing, should we add explicit tests for those scenarios or treat them as default values?
> 
> I’ll update the tests accordingly. Thanks again!

I think we should explicitly test the boundary cases of missing fields, rather than relying solely on default values, to ensure the logic is correct when fields are missing.

**Comment by Ferryx349** [comment]
Sure, I will fix it shortly..
Thanks :)

**Comment by Ferryx349** [comment]
hi @FAUST-BENCHOU @XiShanYongYe-Chang, I have added required changes,
happy to iterate on further feedback..

Thanks

**Comment by XiShanYongYe-Chang** [comment]
Let's wait for @FAUST-BENCHOU to confirm it again.

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interprethealth-test.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  kind: SparkApplication
+  metadata:
+    name: spark-pi
+    namespace: default
+  status:
+    applicationState:
+      state: ""
+operation: InterpretHealth
+output:
+  healthy: true
```
Comment (FAUST-BENCHOU) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interprethealth-test.yaml:63]: SparkApplication unhealthy when applicationState.state is empty
why is the output healthy?

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interpretstatus-test.yaml [hunk 2]
```diff
@@ -75,3 +79,79 @@ observedObj:
...
+  apiVersion: sparkoperator.k8s.io/v1beta2
+  kind: SparkApplication
+  metadata:
+    name: spark-pi
+    namespace: default
+  status:
+    applicationState: {}
+operation: InterpretStatus
+output:
+  status: {}
```
Comment (FAUST-BENCHOU) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interpretstatus-test.yaml:124]: The message is "InterpretStatus should copy applicationState even if the state field is missing".but the expected output is: status: {}. This is inconsistent with the description of "copy applicationState" (if it is to be copied, the applicationState field should be included, even if the map is empty).

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/aggregatestatus-test.yaml [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # test case for aggregating status of SparkApplication
 # case1. SparkApplication with three status items
+# case2. SparkApplication with single status item
+# case3. SparkApplication with no status items
+# case4. SparkApplication with missing attempts fields
+# case5. SparkApplication with missing executorState
+# case6. SparkApplication with missing lastSubmissionAttemptTime
+# case7. SparkApplication with missing terminationTime
```
Comment (FAUST-BENCHOU) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/aggregatestatus-test.yaml:8]: The comments need to be updated; they don't match the test cases.

**Comment by FAUST-BENCHOU** [comment]
@XiShanYongYe-Chang overall great!

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks @Abhay349 @FAUST-BENCHOU 
/lgtm
/approve

**Comment by XiShanYongYe-Chang** [comment]
Hi @Abhay349, just as @FAUST-BENCHOU said, some use case descriptions seem to be inconsistent with the content. Could you help revise the descriptions of the use cases? (The test cases themselves are fine.)

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interprethealth-test.yaml (hunk 1 above)
Comment (Ferryx349) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interprethealth-test.yaml:63]: According to logic,
empty string  "" !="FAILED",so state == ""  is treated as healthy

Comment (FAUST-BENCHOU) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interprethealth-test.yaml:63]: got that,maybe you can change comment to like `SparkApplication healthy when applicationState.state is empty`

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interpretstatus-test.yaml (hunk 2 above)
Comment (Ferryx349) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/sparkoperator.k8s.io/v1beta2/SparkApplication/testdata/interpretstatus-test.yaml:124]: Apologies for description issue,
"InterpretStatus does not copy applicationState when state field is missing", will it work?
